src/ripple/protocol/impl/AccountID.cpp

```diff
@@ -17,17 +17,101 @@
 */
 //==============================================================================
 
+#include <ripple/basics/hardened_hash.h>
 #include <ripple/protocol/AccountID.h>
 #include <ripple/protocol/PublicKey.h>
 #include <ripple/protocol/digest.h>
 #include <ripple/protocol/tokens.h>
+#include <array>
 #include <cstring>
+#include <mutex>
 
 namespace ripple {
 
+namespace detail {
+
+/** Caches the base58 representations of AccountIDs */
+class AccountIdCache
+{
+private:
+    // The total number of items the cache buffer can contain
+    std::size_t const count_;
+
+    // The maximum size (in bytes) of an item, along with a null terminator
+    // and aligned at 8 bytes
+    std::size_t const size_ = 64;
+
+    // The offset of the string representation of an AccountID
+    std::size_t const offset_ = 24;
+
+    // We use a hash function designed to resist algorithmic complexity attacks
+    hardened_hash<> hasher_;
+
+    // The memory block
+    std::unique_ptr<char[]> chars_;
+
+    std::array<std::mutex, 16> mutex_;
+
+public:
+    AccountIdCache(std::size_t count)
+        : count_(count), chars_(std::make_unique<char[]>(count_ * size_))
+    {
+        std::memset(chars_.get(), 0, count_ * size_);
+    }
+
+    std::string
+    toBase58(AccountID const& id)
+    {
+        assert(chars_.get() != nullptr);
+
+        auto const index = hasher_(id.data()) % count_;
+
+        auto buf = chars_.get() + (size_ * index);
+
+        assert(
+            buf != nullptr && buf >= chars_.get() &&
+            buf < (chars_.get() + (count_ * size_)));
+
+        {
+            std::lock_guard lock(mutex_[index % mutex_.size()]);
+
+            if (buf[offset_] != 0 &&
+                std::memcmp(buf, id.data(), id.size()) == 0)
+                return buf;
```

Comment:
I think it should be: `return buf + offset;`

Reply:
Fixed (code no longer there)